rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+responds with a `200 OK` as soon as the request is received. This `200 OK` does
+not guarantee that the Splunk event data in the request is persisted or fully
+processed.
+
+With indexer acknowledgement enabled, the `200 OK` response contains an
+additional JSON body field specifying an `ackId` where `ackId` is an integer
+identifier corresponding to the request. Note that a single HEC request can
+include multiple Splunk events, so the `ackId` is per-request.
+
+```{}
+// Example response from local Splunk 7.3.2
+{
+   "text": "Success",
+   "code": 0,
+   "ackId": 10
+}
+```
+
+Using one or more `ackId`’s, the user can query Splunk’s
+`/services/collector/ack` endpoint to check on the status of associated
+requests. For each `ackId`, Splunk returns `true`/`false` depending on whether
+the data in the request has been persisted (in Splunk’s words, “replicated at
+the desired replication factor”). Upon returning `true`, Splunk drops the
+`ackId` status and all subsequent requests with the same `ackId` will return
+`false` (based on Splunk channel-related expiration settings, `ackId`'s can be
+reset/reused).
+
+```{}
+// Example request body
+{
+    "acks": [0, 1, 2]
+}
+
+// Example response body
+{
+    "acks":
+        {
+            "0": true,
+            "1": false,
+            "2": true
+        }
+}
+```
+
+In addition to the overall protocol, there are a few details worth highlighting:
+
+* Authentication tokens
+  * To submit requests to HEC in general, an authentication token must be
+    included in request headers. Indexer acknowledgement is enabled per
+    authentication token in Splunk. The `splunk_hec` source currently supports a
+    list of valid authentication tokens. The `splunk_hec` sinks currently
+    support configuring a single authentication token.
+
+* Splunk channels
+  * Upon enabling indexer acknowledgement, Splunk requires HEC requests to
+    include a channel ID (any valid `guid` value). Querying
+    `/services/collector/ack` also requires a channel ID.
+  * `ackId`’s are assigned per-channel.
+
+* Pending acks
+  * Clients are not required or guaranteed to query for `ackId` status. To avoid
+    running out of memory from `ackId` build up, Splunk offers a few
+    configuration options to limit the number of pending `ackId`’s both overall
+    and per-channel.
+  * Splunk does not explicitly indicate whether event data is dropped or still
+    pending processing. Either of these states will result in `{ack_id}: false`
+    status. Splunk only advises that after a certain amount of time (e.g. 5
+    minutes), the data can be considered dropped.
+
+## Proposal: `splunk_hec` Source Indexer Acknowledgement
+
+### User Experience
+
+Users can configure the `splunk_hec` source with additional indexer acknowledgement related settings.
+
+```{toml}
+[sources.splunk_hec]
+type = “splunk_hec”
+...
+acknowledgements.enabled = true
+acknowledgements.max_pending_acks = 1_000_000
```

Comment:
Sure, I can see this simply being `acknowledgements` boolean and `max_pending_acks`. When validating configs, is it possible to check that `acknowledgements` has been set as a pre-requisite for setting `max_pending_acks`?

Reply:
It seems like there are three states:

1. Indexer acknowledgements not enabled
2. Indexer acknowledgements enabled, but end-to-end acknowledgements in Vector not enabled
3. Indexer acknowledgements enabled and wired into Vector end-to-end acknowledgements

I don't think 2 is as useful to model so we could wait for a user to ask for it. It'd really just be useful for users that want 1, but whatever they are configuring to send to Vector requires it. I could see modeling this as an extra boolean flag on `acknowledgements` if we do decide to add it later.

I wonder if we could update acknowledgements everywhere to, rather than just being a boolean, be an object with an `enabled` key. This would allow the additional configuration here and open the door for other source-specific acknowledgement configuration.

Reply:
I am certainly open to rewriting the acking config to use `acknowledgements.enabled`. I think we have precedent for options that take either a structure or bool for backwards compatibility.

Reply:
Done in #9883 although I'm not sure it entirely helps here, as it's a global structure and can't be easily added to for this source.

Reply:
Mmm, yeah, it does seem like we'll need to have sources that need to extend it implement their own `AcknowledgementsConfig`.